### DISCUSSION TITLE
How can I connect to WSL2 distro?

### DISCUSSION THREAD

**Discussion by palcoo42** [eclipse-theia/theia#13857] [Q&A]
Hello,

I have installed Theia IDE in Windows and I have tried to connect to my existing WSL2 distribution. However I did not found an option for this in "Open a Remote Window". Asfaik this option is provided by Microsoft extension in VS Code and I could not found it in Theia extensions. Is there a option how can I connect to WSL2 from Windows Theia IDE installation?

Thanks in advance

**Comment by palcoo42** [comment]
Print screen of "Open a Remote Window"

![image](https://github.com/eclipse-theia/theia/assets/122668559/178812aa-ae9e-4ea3-adfb-50dbb2afe442)

**Reply by dscain** [reply]
Thanks for the information. How to configure a private key? Usually I just add all to my .ssh\config, but it seems it is not being picked up by Theia. Also, I cannot see additional debugging information in the Output window. I am using Theia 1.51.0 installed on Windows 11 trying to connect to Ubuntu machines. Thanks in advance for the support!

**Comment by msujew** [comment]
Hey @palcoo42,

Theia currently only supports to work on a remote machine via SSH or to run your workspace in a dev-container. WSL2 is not supported yet.

Each of those features is fairly complicated from a technical perspective and none of the contributing parties have communicated interest in supporting WSL specificially.

**Reply by palcoo42** [reply]
Thanks for a reply, unfortunately WSL2 is a must have for me, so for now I am stick with vs-code

**Comment by nickger** [comment]
Following [this](https://github.com/eclipse-theia/theia/pull/12618) I could run the command "connect to host" with my WSL credentials, it did something for quite a while but then opened a Theia instance on the host machine. @msujew is WSL still not supported even through SSH?

**Reply by msujew** [reply]
If it's just using SSH, it's definitely supported. We just don't have dedicated WSL extension like VS Code does.

**Reply by nickger** [reply]
Yes, it does just use SSH which works fine in Cursor via VS Code extension. But it opens a new instance of Theia on the host machine.

Though, it does indtall something - there are folders ".theia-ide-electron-app-1.59.1-remote" and ".theia-ide" in user's home. What can be wrong?

**Reply by msujew** [reply]
Similar to Cursor/VS Code, Theia needs to reinstall itself onto the host machine with the correct native binaries in order to run on a different OS.  The `.theia-ide-electron-app-1.59.1-remote` contains the server that is required to make the remote feature work, whereas `.theia-ide` contains all app-wide settings, etc.

**Reply by nickger** [reply]
Great. How make it work?

**Reply by msujew** [reply]
I thought it works? You mentioned:

> then opened a Theia instance on the host machine.

I assumed that was a confirmation that it works. Do you get an error or something similar?

**Reply by nickger** [reply]
Sorry to confuse you, but newly opened instance does not see anything on the remote machine and I cannot start working on the remotely hosted code. May be I don't see something - this feature is not documented. In VS Code I am still in the same VS Code instance and work as the code is on my machine. The only difference are extensions: I clearly see what is installed on remote machine and what - on host. What should I see/do in Theia..?

**Reply by msujew** [reply]
@jonah-iden Any clue?

**Reply by jonah-iden** [reply]
sadly not, i need to test this myself. I don't fully understand what the error currently is

**Reply by nickger** [reply]
What I consider as an error is: newly opened instance of Theia does not see files on the remote machine. So, I cannot open remote folder and start remote development.

**Reply by jonah-iden** [reply]
Ok thanks for the info. Is theia showing it is connected to a remote server in the bottom left? Because if you can not see the filesm it seems like it didn't connect correctly

**Reply by nickger** [reply]
![Capture](https://github.com/user-attachments/assets/517819b2-9733-4797-a814-badd0eeb0111)
This is what I see at the end: 2 identical instances of Theia.

**Reply by jonah-iden** [reply]
ok thanks. Seems i was correct, its not connecting properly to the server in wsl. I'll have to test this myself

**Reply by MuratOzsoyler** [reply]
Theia can connect to WSL2 via SSL using password authentication, but can not using public key. `ssh-agent` is working and I can connect in Powershell using `ssh` command. Am I missing something?

**Reply by jonah-iden** [reply]
I havn't gotten around to testing this in depth. It just seems that wsl ssh connections do have a lot of problem curently

**Reply by jonah-iden** [reply]
I started implementing native WSL support for theias remote feature. Here is the [draft Pr](https://github.com/eclipse-theia/theia/pull/15543) . Feel free to test it out if you have time. It's very basic still and has some issues but its a start.

**Reply by jonah-iden** [reply]
Btw `connect current window to host` should wokr. It seems there is currently an issue with ports not being set correctly when opening a new window to connect to an ssh server

**Reply by nickger** [reply]
@jonah-iden How can I try what you've done? Do I need to download the source code and compile? I am not familiar with npm and such...

**Reply by jonah-iden** [reply]
unfortunately yes you would have to build it from source. 
You would need to have node (best would be probably version 20) installed. 
Then just clone the branch from the PR, exeute `npm i` and then npm `npm run build:electron` then you should be able to start the application with `npm run start:electron`

**Comment by MuratOzsoyler** [comment]
"connect current window to host should wokr"
Unfortunately result is same: Could not open SSH connection. All configured
authentication methods failed.

**Reply by jonah-iden** [reply]
sorry i should have specified this was about nickger issue. Havn't tested the public key one yet

**Reply by jonah-iden** [reply]
Are you sure you setup SSH correctly in your WSL instance? I just tested it locally with my public key in the `authorized_keys` in my ubuntu distro and it worked fine for me

**Comment by MuratOzsoyler** [comment]
I can connect from PowerShell CLI using ssh command. ssh-agent is running
in Windows side and my private key is ssh-added. My Windows and WSL
Archlinux instance is updated recently so the software versions should be
the latest.

Could it be a configuration issue? Can I get SSH connection log of Theia?

**Reply by jonah-iden** [reply]
You could try jjust launching theia from a terminal. That should give you at least some output. There is not a lot of logging built into the ssh feature. what you could try would be to build theia from source and debug it yourself. I could imagine you are not using a default format for your key maybe? Default currently include
```
        const PATH_SSH_CLIENT_ID_DSA = path.join(homeDir, '.ssh', '/id_dsa');
        const PATH_SSH_CLIENT_ID_ECDSA = path.join(homeDir, '.ssh', '/id_ecdsa');
        const PATH_SSH_CLIENT_ID_RSA = path.join(homeDir, '.ssh', '/id_rsa');
        const PATH_SSH_CLIENT_ID_ED25519 = path.join(homeDir, '.ssh', '/id_ed25519');
        const PATH_SSH_CLIENT_ID_XMSS = path.join(homeDir, '.ssh', '/id_xmss');
        const PATH_SSH_CLIENT_ID_ECDSA_SK = path.join(homeDir, '.ssh', '/id_ecdsa_sk');
        const PATH_SSH_CLIENT_ID_ED25519_SK = path.join(homeDir, '.ssh', '/id_ed25519_sk');
``` 
or the key has to be in your ssh config file. But i actually don't see the parsing in the code right now. Maybe thats acutally a still missing feature

**Reply by jonah-iden** [reply]
should you try to debug it yourself, take a look at the `packages\remote\src\electron-node\ssh\remote-ssh-connection-provider.ts:142` file, specifcly the `establishSSHConnection()´ method

**Comment by jonah-iden** [comment]
Btw the [WSL support](https://github.com/eclipse-theia/theia/pull/15543) is working now and should hopefully make it into the next release
